main/modes/pinball/mode_pinball.c

```diff
@@ -0,0 +1,149 @@
+static void pinBackgroundDrawCallback(int16_t x, int16_t y, int16_t w, int16_t h, int16_t up, int16_t upNum);
+
+//==============================================================================
+// Variables
+//==============================================================================
+
+const char pinballName[] = "Pinball";
+
+swadgeMode_t pinballMode = {
+    .modeName                 = pinballName,
+    .wifiMode                 = NO_WIFI,
+    .overrideUsb              = false,
+    .usesAccelerometer        = false,
+    .usesThermometer          = false,
+    .overrideSelectBtn        = false,
+    .fnEnterMode              = pinEnterMode,
+    .fnExitMode               = pinExitMode,
+    .fnMainLoop               = pinMainLoop,
+    .fnAudioCallback          = NULL,
+    .fnBackgroundDrawCallback = pinBackgroundDrawCallback,
+    .fnEspNowRecvCb           = NULL,
+    .fnEspNowSendCb           = NULL,
+    .fnAdvancedUSB            = NULL,
+};
+
+pinball_t* pinball;
+
+//==============================================================================
+// Functions
+//==============================================================================
+
+/**
+ * @brief Set up the pinball mode
+ */
+static void pinEnterMode(void)
+{
+    // Make it fast
+    setFrameRateUs(PIN_US_PER_FRAME);
+
+    // Allocate all the memory
+    pinball = calloc(sizeof(pinball_t), 1);
+
+    pinball->balls   = heap_caps_calloc(MAX_NUM_BALLS, sizeof(pbCircle_t), MALLOC_CAP_SPIRAM);
+    pinball->bumpers = heap_caps_calloc(MAX_NUM_BUMPERS, sizeof(pbCircle_t), MALLOC_CAP_SPIRAM);
+    pinball->walls   = heap_caps_calloc(MAX_NUM_WALLS, sizeof(pbLine_t), MALLOC_CAP_SPIRAM);
+
+    pinball->ballsTouching = heap_caps_calloc(MAX_NUM_BALLS, sizeof(pbTouchRef_t*), MALLOC_CAP_SPIRAM);
+    for (uint32_t i = 0; i < MAX_NUM_BALLS; i++)
+    {
+        pinball->ballsTouching[i] = heap_caps_calloc(MAX_NUM_TOUCHES, sizeof(pbTouchRef_t), MALLOC_CAP_SPIRAM);
+    }
+
+    // Split the table into zones
+    createTableZones(pinball);
+
+    // Create random balls
+    createRandomBalls(pinball, 3);
+
+    // Create random walls
+    createRandomWalls(pinball, 100);
+
+    // Create random bumpers
+    createRandomBumpers(pinball, 10);
+
+    // Load font
+    loadFont("ibm_vga8.font", &pinball->ibm_vga8, false);
+}
+
+/**
+ * @brief Tear down the pinball mode
+ *
+ */
+static void pinExitMode(void)
+{
+    for (uint32_t i = 0; i < MAX_NUM_BALLS; i++)
+    {
+        free(pinball->ballsTouching[i]);
+    }
+    free(pinball->ballsTouching);
+
+    free(pinball->balls);
+    free(pinball->walls);
+    free(pinball->bumpers);
+    // Free font
+    freeFont(&pinball->ibm_vga8);
+    // Free the rest of the state
+    free(pinball);
+}
+
+/**
+ * @brief Run the pinball main loop
+ *
+ * @param elapsedUs The number of microseconds since this function was last called
+ */
+static void pinMainLoop(int64_t elapsedUs)
+{
+    // Make a local copy for speed
+    pinball_t* p = pinball;
+
```

Comment:
You can't exit the mode since it never checks the button queue, but I assume "add controls" is pretty high on the todo list

Reply:
Yep. There's a lot to do there...